### PR TITLE
Automatic installation of eslint language server

### PR DESCRIPTION
**Pull request by yyoncho**
- Fixed also handling of `eslint/confirmLocalESLint`

----

#

### CODE REVIEW

File: lsp-mode.el [hunk 1]
```diff
@@ -7033,6 +7033,46 @@ STORE-PATH to make it executable."
...
+                                  (t nil))
+  "The script to unzip."
+  :group 'lsp-mode
+  :type 'string
+  :package-version '(lsp-mode . "7.1"))
+
+(defun lsp-unzip (zip-file dest)
+  "Get extension from URL and extract to DEST."
+  (unless lsp-unzip-script
+    (error "Unable to find `unzip' or `powershell' on the path."))
```
Comment (kiennq) [inline lsp-mode.el:7057]: Error should not end with `.`
```suggestion
    (error "Unable to find `unzip' or `powershell' on the path, please customize `lsp-unzip-script'"))
```

File: clients/lsp-eslint.el [hunk 2]
```diff
@@ -36,6 +36,18 @@
   :group 'lsp-mode
   :link '(url-link "https://github.com/Microsoft/vscode-eslint"))
 
+(defcustom lsp-eslint-zip-path (f-join lsp-server-install-dir "eslint/extension.zip")
+  "The path to the file in which `rust-analyzer' will be stored."
```
Comment (ericdallo) [inline clients/lsp-eslint.el:40]: wrong docstring?

File: clients/lsp-eslint.el [hunk 3]
```diff
@@ -36,6 +36,18 @@
...
+  :type 'file
+  :group 'lsp-eslint
+  :package-version '(lsp-mode . "7.1"))
+
+(defcustom lsp-eslint-unziped-path (f-join lsp-server-install-dir "eslint/unziped")
+  "The path to the file in which `rust-analyzer' will be stored."
```
Comment (ericdallo) [inline clients/lsp-eslint.el:46]: wrong docstring?

File: clients/lsp-eslint.el [hunk 4]
```diff
@@ -36,6 +36,18 @@
   :group 'lsp-mode
   :link '(url-link "https://github.com/Microsoft/vscode-eslint"))
 
+(defcustom lsp-eslint-zip-path (f-join lsp-server-install-dir "eslint/extension.zip")
```
Comment (kiennq) [inline clients/lsp-eslint.el:39]: Is this needed? I think we should just keep the `zip` file as temp file. There's no value customizing it.

File: clients/lsp-eslint.el (hunk 3 above)
Comment (kiennq) [inline clients/lsp-eslint.el:46]:
```suggestion
  "The path to the file in which `eslint' will be stored."
```

File: clients/lsp-eslint.el (hunk 4 above)
Comment (yyoncho) [reply clients/lsp-eslint.el:39]: It can go away, I will update the PR.

File: clients/lsp-eslint.el (hunk 2 above)
Comment (yyoncho) [reply clients/lsp-eslint.el:40]: fixed.

File: clients/lsp-eslint.el (hunk 4 above)
Comment (yyoncho) [reply clients/lsp-eslint.el]: Fixed now.

File: lsp-mode.el [hunk 5]
```diff
@@ -7033,6 +7033,46 @@ STORE-PATH to make it executable."
       (set-file-modes store-path #o0700)
       store-path)
      ((f-exists? store-path) store-path))))
+
```
Comment (wyuenho) [inline lsp-mode.el:7036]: Maybe start moving new util functions to a new lsp-utils.el file as suggested in #2167 ?

Comment (yyoncho) [reply lsp-mode.el:7036]: I don't see this as a top priority issue and I don't want to spend time on it. I don't mind if someone wants to take on it.

File: clients/lsp-eslint.el [hunk 6]
```diff
@@ -36,6 +36,12 @@
   :group 'lsp-mode
   :link '(url-link "https://github.com/Microsoft/vscode-eslint"))
 
+(defcustom lsp-eslint-unziped-path (f-join lsp-server-install-dir "eslint/unziped")
```
Comment (leungbk) [inline clients/lsp-eslint.el:39]:
```suggestion
(defcustom lsp-eslint-unzipped-path (f-join lsp-server-install-dir "eslint/unzipped")
```

File: clients/lsp-eslint.el [hunk 7]
```diff
@@ -227,16 +238,32 @@ source.fixAll code action."
      (when (called-interactively-p 'any)
        (lsp--info "source.fixAll.eslint action not available")))))
 
+(defun lsp-eslint-server-command ()
+  (if (lsp-eslint-server-exists? lsp-eslint-server-command)
+      lsp-eslint-server-command
+    `(,lsp-eslint-node ,(f-join lsp-eslint-unziped-path
```
Comment (leungbk) [inline clients/lsp-eslint.el:244]:
```suggestion
    `(,lsp-eslint-node ,(f-join lsp-eslint-unzipped-path
```

File: clients/lsp-eslint.el [hunk 8]
```diff
@@ -263,7 +291,20 @@ source.fixAll code action."
...
-                                                :glob-pattern "**/package.json")])))))))
+                                                :glob-pattern "**/package.json")])))))
+  :download-server-fn (lambda (_client callback error-callback _update?)
+                        (let ((tmp-zip (make-temp-file "ext" nil ".zip")))
+                          (delete-file tmp-zip)
+                          (lsp-download-install
+                           (lambda (&rest _)
+                             (condition-case err
+                                 (progn
+                                   (lsp-unzip tmp-zip lsp-eslint-unziped-path)
```
Comment (leungbk) [inline clients/lsp-eslint.el:302]:
```suggestion
                                   (lsp-unzip tmp-zip lsp-eslint-unzipped-path)
```

**Review by kiennq — APPROVED** [review]

**Review by ericdallo — APPROVED** [review]

**Pull request merged**